src/util/typing.js

```diff
@@ -0,0 +1,20 @@
+/**
+ * Helper function for cases in which we know it's safe to apply a Ref
+ * that is a different Element type than the particular element expects.
+ *
+ * For example, the following will cause a type-checking error:
+ *
+ * param {Object} props
+ *   prop {import('preact').Ref<HTMLElement>} props.myRef
+ *
+ * function MyThing({ myRef }) {
+ *   return <div ref={myRef}>Hi there</div>;
+ * }
+ *
+ * The <div> wants {Ref<HTMLDivElement>}, while the component accepts the more
+ * generic {Ref<HTMLElement>}. TS' invariance in this case gets in our way.
+ * We want to keep the flexibility of accepting an {Ref<HTMLElement>} here.
+ */
+export function refElement(arg) {
+  return /** @type {import('preact').Ref<any>} */ (arg);
+}
```

Comment:
This explanation is clear although I think the term "downcast" probably best describes what we're doing with the input. I took a shot at creating a generic version of this that has the proper types:

```js
/**
 * @template T
 * @typedef {import('preact').Ref<T>} Ref
 */

/**
 * Helper for downcasting a ref to a more specific type, where that is safe
 * to do.
 *
 * This is mainly useful to cast a generic `Ref<HTMLElement>` to a more specific
 * element type (eg. `Ref<HTMLDivElement>`) for use with the `ref` prop of a JSX element.
 * Since Preact only writes to the `ref` prop, such a cast is safe.
 *
 * @template T
 * @template {T} U
 * @param {Ref<T>|undefined} ref
 * @return {Ref<U>|undefined}
 */
export function downcastRef(ref) {
  return /** @type {Ref<U>|undefined} */ (ref);
}
```

The above works in place of the current `refElement` calls, if you adjust the name. The type variables T and U could also be given more descriptive names like `Base` and `Derived`.

Reply:
> This explanation is clear although I think the term "downcast" probably best describes what we're doing with the input.

Heh, this is why I specifically asked for vocabulary when we chatted about this. I'll use your suggestions here.